howso/train_ts_ablation.amlg

```diff
@@ -0,0 +1,562 @@
+						(if one_id
+							(zip (map
+								(lambda (get (current_value) id_index) )
+								data
+							))
+						)
+					previous_id (null)
+				)
+
+				(while (< (current_index) (size data))
+					;first row sets previous_id and unique_ids_set into (previous_result)
+					(if (= 0 (current_index))
+						(seq
+							(assign (assoc
+								previous_id (get data [(current_index 2) id_index])
+							))
+							(remove unique_ids_set previous_id)
+						)
+
+						;if the current id doesn't match previous_id, check to see if it has
+						;already been encountered. if so, the data is out of order
+						(!= previous_id (get data [(current_index 1) id_index]))
+						(seq
+							(assign (assoc
+								previous_id (get data [(current_index 2) id_index])
+							))
+
+							;new ids should still be in unique_ids_set, but if they are not they've been encountered before
+							(if (not (contains_index (previous_result 0 (true)) previous_id))
+								(conclude (conclude
+									(assign (assoc needs_grouping (true) ))
+								))
+							)
+
+							;id has not been encountered before, remove it from unique_ids_set
+							(remove (previous_result) previous_id)
+						)
+
+						;else no change to unique_ids_set
+						(previous_result)
+					)
+				)
+			)
+
+			;else multiple ids, less efficient
+			(let
+				(assoc
+					unique_ids_list
+						(values
+							(map (lambda (unzip (current_value) id_indices)) data)
+							(true)
+						)
+					previous_ids []
+				)
+
+				(while (< (current_index) (size data))
+					;first row sets previous_id and unique_ids_list into (previous_result)
+					(if (= 0 (current_index))
+						(seq
+							(assign (assoc
+								previous_ids (unzip (get data (current_index 1)) id_indices)
+							))
+							(filter
+								(lambda (!= (current_value) previous_ids))
+								unique_ids_list
+							)
+						)
+
+						;if the current ids don't match previous_ids, check to see if they have
+						;already been encountered. if so, the data is out of order
+						(!= previous_ids (unzip (get data (current_index)) id_indices) )
+						(seq
+							(assign (assoc
+								previous_ids (unzip (get data (current_index 1)) id_indices)
+							))
+
+							;new ids should still be in unique_ids_list, but if they are not they've been encountered before
+							(if (not (contains_value (previous_result 0 (true)) previous_ids))
+								(conclude (conclude
+									(assign (assoc needs_grouping (true) ))
+								))
+							)
+
+							;id has not been encountered before, remove it from unique_ids_list
+							(filter
+								(lambda (!= (current_value) previous_ids))
+								(previous_result)
+							)
+						)
+
+						;else no change to unique_ids_list
+						(previous_result)
+					)
+				)
+			)
+		)
+
+		;output true if data needs grouping
+		needs_grouping
+	)
+
+
+	;hepler method
+	;uses cases and id_indices to output cases grouped by ids in columns specified by id_indices
+	#!GroupDataByIds
```

Comment:
is this really better than just a multi column sort?

Reply:
Well, it's grouping without sorting, because it's not comparing the values to see which should go first or second, simply arranging them so they are lumped together in whatever order they happen to be. not sure it's better or worse, we can discuss.

Reply:
if we sort, it will actually sort the groups by ids, whereas current code leaves the groups in the same order as they are, so if there are minimal changes that need to happen, my code will disturb it the least. not sure if it's slower than sort.